Applications/Spire/Source/Ui/RealSpinBox.cpp

```diff
@@ -0,0 +1,441 @@
+        lineEdit()->selectedText());
+      current_text.insert(lineEdit()->cursorPosition(), input);
+      if(is_valid(current_text.trimmed())) {
+        QAbstractSpinBox::keyPressEvent(event);
+      }
+    }
+  }
+}
+
+void RealSpinBox::mouseMoveEvent(QMouseEvent* event) {
+  auto control = get_current_control(event->pos());
+  if(m_up_arrow_timer_id != -1 &&
+      control != QStyle::SubControl::SC_ScrollBarAddLine) {
+    stop_timer(m_up_arrow_timer_id);
+  } else if(m_down_arrow_timer_id != -1 &&
+      control != QStyle::SubControl::SC_ScrollBarSubLine) {
+    stop_timer(m_down_arrow_timer_id);
+  }
+  QAbstractSpinBox::mouseMoveEvent(event);
+}
+
+void RealSpinBox::mousePressEvent(QMouseEvent* event) {
+  if(event->button() == Qt::LeftButton && !m_has_first_click) {
+    selectAll();
+    m_has_first_click = true;
+  }
+  auto control = get_current_control(event->pos());
+  if(control == QStyle::SubControl::SC_ScrollBarAddLine) {
+    m_up_arrow_timer_id = startTimer(qApp->keyboardInputInterval());
+  } else if(control == QStyle::SubControl::SC_ScrollBarSubLine) {
+    m_down_arrow_timer_id = startTimer(qApp->keyboardInputInterval());
+  }
+  QAbstractSpinBox::mousePressEvent(event);
+}
+
+void RealSpinBox::mouseReleaseEvent(QMouseEvent* event) {
+  stop_timer(m_up_arrow_timer_id);
+  stop_timer(m_down_arrow_timer_id);
+  QAbstractSpinBox::mouseReleaseEvent(event);
+}
+
+void RealSpinBox::timerEvent(QTimerEvent* event) {
+  if(event->timerId() == m_up_arrow_timer_id) {
+    stop_timer(m_up_arrow_timer_id);
+    m_up_arrow_timer_id = startTimer(MOUSE_REPEAT_DELAY_MS);
+    add_step(1, qApp->keyboardModifiers());
+  } else if(event->timerId() == m_down_arrow_timer_id) {
+    stop_timer(m_down_arrow_timer_id);
+    m_down_arrow_timer_id = startTimer(MOUSE_REPEAT_DELAY_MS);
+    add_step(-1, qApp->keyboardModifiers());
+  }
+}
+
+QAbstractSpinBox::StepEnabled RealSpinBox::stepEnabled() const {
+  return QAbstractSpinBox::StepUpEnabled | QAbstractSpinBox::StepDownEnabled;
+}
+
+void RealSpinBox::set_decimal_places(int decimals) {
+  m_decimals = std::min(decimals, MAXIMUM_DECIMAL_PLACES);
+  auto point_count = [&] {
+    if(m_decimals > 0) {
+      return 1;
+    }
+    return 0;
+  }();
+  m_real_regex = QRegularExpression(QString(
+    "^[%1]?([0-9]*[%2]{0,%3}[0-9]{0,%4})$").arg(m_locale.negativeSign())
+    .arg(m_locale.decimalPoint()).arg(point_count).arg(m_decimals));
+}
+
+void RealSpinBox::set_minimum_decimal_places(int decimals) {
+  m_minimum_decimals = decimals;
+}
+
+void RealSpinBox::set_minimum(Real minimum) {
+  m_minimum = minimum;
+  auto value = get_value(text());
+  if(value != boost::none && value->compare(m_minimum) < 0) {
+    m_last_valid_value = m_minimum;
+    lineEdit()->setText(display_string(m_minimum));
+  }
+}
+
+void RealSpinBox::set_maximum(Real maximum) {
+  m_maximum = maximum;
+  auto value = get_value(text());
+  if(value != boost::none && value->compare(m_maximum) > 0) {
```

Comment:
See above.

Reply:
Is there a syntax for doing something like:

`auto value = get_value(text()) && value->compare(m_maximum) > 0`

Wrapping the assignment in parentheses makes the compiler think it's an incorrect cast to auto, and not differentiating the assignment makes this statement incorrect because value isn't assigned before being evaluated.

Reply:
Yes:

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0305r0.html